narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,233 +3,356 @@
                 }
-                () = &mut timer => {
-                    // We still would like to GC even if we do not receive anything from either the synchronizer or
-                    //  the Waiter. This can happen, as we may have asked for certificates that (at the time of ask)
-                    // were not past our GC bound, but which round swept up past our GC bound as we advanced rounds
-                    // & caught up with the network.
-                    // Those certificates may get stuck in pending, unless we periodically clean up.
-
-                    // Reschedule the timer.
-                    timer.as_mut().reset(Instant::now() + Duration::from_millis(GC_RESOLUTION));
-                    attempt_garbage_collection = true;
-                },
-
-                Ok(()) = self.rx_consensus_round_updates.changed() => {
-                    attempt_garbage_collection = true;
-                }
-
             }
+        }
+    }
 
-            // Either upon time-out or round change
-            if attempt_garbage_collection {
-                let round = *self.rx_consensus_round_updates.borrow();
-                if round > self.gc_depth {
-                    let gc_round = round - self.gc_depth;
-
-                    self.pending.retain(|_digest, (r, once_cancel)| {
-                        if *r <= gc_round {
-                            // note: this send can fail, harmlessly, if the certificate has been delivered (`notify_read`)
-                            // and the present code path fires before the corresponding `waiting` item is unpacked above.
-                            let _ = once_cancel
-                                .take()
-                                .expect("This should be protected by a write lock")
-                                .send(());
-                            false
-                        } else {
-                            true
-                        }
-                    });
-                }
+    // Starts a task to fetch missing certificates from other primaries.
+    // A call to kick() can be triggered by a certificate with missing parents or the end of a fetch task.
+    // Each iterations of kick() updates the target rounds, and iterations will continue until there is no more target rounds to catch up to.
+    #[allow(clippy::mutable_key_type)]
+    fn kick(&mut self) {
+        let progression = match self.read_round_progression() {
+            Ok(progression) => progression,
+            Err(e) => {
+                warn!("Failed to read rounds per authority from the certificate store: {e}");
+                return;
             }
+        };
+        self.targets.retain(|origin, target_round| {
+            let current_round = progression.get(origin).unwrap();
+            // Drop sync target when cert store already has an equal or higher round for the origin.
+            current_round < target_round
+        });
```

Comment:
Good point. I will rework the approach in this PR.

Reply:
After https://github.com/MystenLabs/sui/pull/5238, committing a certificate earlier than the last committed round of the same validator becomes disallowed in bullshark. So in internal consensus, the case above can only happen for uncommitted certificates.
- For internal consensus, missing certificates will be fetched starting from the last committed rounds, instead of last written rounds to avoid the issues above.
- For external consensus, I wonder if we can disable the case above, and continue using last written rounds for the starting point of fetching missing certificates.

I also considered disallowing the case above for internal consensus as well, but I cannot convince myself it is safe.

Reply:
@mwtian shall we catch up about this? I am a bit confused on the internal vs external consensus. To be honest I would try to make as few assumptions as possible here. I would keep this component agnostic of the consensus protocol functionality.